File: stdlib5/kweb/kweb.py
# test561 : kweb (py)

import requests
import bs4

# http~.html에서 domain 별 txt 반환
def gettxt(url, domain):
    html = requests.get(url).text
    dom = bs4.BeautifulSoup(html, "html.parser")
    data = dom.find("p", id = domain)
    if data == None:
        raise Exception(f"No Domain : {domain}")
    else:
        return data.text

# http~/ + name + .num 를 path로 바이너리 생성
def download(url, name, num, path):
    if url[-1] != "/":
        url = url + "/"
    with open(path, "wb") as f:
        for i in range(0, num):
            html = requests.get(url + name + f".{i}")
            if html.content == None:
                raise Exception(f"No Data : {name}.{i}")
            else:
                f.write(html.content)

# edgedriver 홈페이지에서 정보 받아오기, 홈페이지 + 행동정보 + zip경로 -> 버전 정수
def driver(url, datas, path):
    # html 요소, 공통 레이어 class 명, html 요소, 버튼 class 명, html 요소, 버전 class 명
    # ["div", "common-card-list__card", "a", "common-button.common-button--tag", "div", "block-web-driver__versions"]
    html = requests.get(url).text
    dom = bs4.BeautifulSoup(html, "html.parser")
    cand = dom.findAll( datas[0], { "class" : datas[1] } )
    box = None
    for i in cand:
        if "stable" in i.text.lower().replace(" ", ""):
            box = i
    if box == None:
        raise Exception("No Stable WD")
    but = None
    cand = box.findAll( datas[2], { "class" : datas[3] } )
    for i in cand:
        if "x64" in i.text.lower().replace(" ", ""):
            but = i
    if but == None:
        raise Exception("No x64 WD")
    link = but.attrs["href"] # x64 driver link
    cand = box.find( datas[4], { "class" : datas[5] } )
    if cand == None:
        ver = 0
    else:
        ver = ""
        num = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]
        for i in cand.text:
            if i in num:
                ver = ver + i
            elif i == "." and num != "":
                break
        ver = int(ver) # 버전 정수
    with open(path, "wb") as f:
        html = requests.get(link)
        if html.content == None:
            raise Exception(f"No Data : {link}")
        else:
            f.write(html.content)
    return ver

# CM이 관리하는 사용자 영역은 extension과 common
# binary와 value는 Starter5와 CM이 자동 관리.
# extension : CM 제외 extension
# common : webdriver 제외 공통 프로그램
# binary : webdriver, CM
# value : MS web data, values -> CM이 common에 자동 업로드
"""
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>kweb datasvr 0</title>
</head>
<body>
    <p id="extension">
    num = 1;<br>
    0.num = 3; 0.name = "gen5win_reltest.zip"; 0.txt = "윈도우용 테스트 프로그램";<br>
    </p>

    <p id="common">
    num = 1;<br>
    0.num = 3; 0.name = "gen5win_reltest.zip"; 0.txt = "윈도우용 테스트 프로그램";<br>
    </p>

    <p id="binary">
    num = 1;<br>
    0.num = 1; 0.name = "msedgedriver.exe"; 0.version = 120.0; 0.data = 20240112;<br>
    </p>
    
    <p id="value">
    name = "Ko";
    <br>
    age = 21;
    </p>

    <p id="explain">
    extension과 common은 content manager용.<br>
    CM에서 binary와 value 데이터 업데이트 가능.<br>
    특별히 webdriver는 MS 홈페이지와 binary로 자동 업데이트.<br>
    binary에 CM은 Starter5용.
    </p>
    
</body>
</html>
"""

File: stdlib5/kweb/kweb.go
package kweb

import (
	"fmt"
	"io"
	"net/http"
	"os"
	"strconv"
	"strings"

	"github.com/PuerkitoBio/goquery"
)

// test562 : kweb (go)

// go mod init example.com
// go get github.com/PuerkitoBio/goquery

// http~.html에서 domain 별 txt 반환
func Gettxt(url string, domain string) string {
	// HTTP GET 요청
	resp, err := http.Get(url)
	if err != nil {
		panic("get fail")
	}
	defer resp.Body.Close()

	// HTTP 응답이 200 OK인지 확인
	if resp.StatusCode != http.StatusOK {
		panic("200 fail")
	}

	// HTML 파싱
	doc, err := goquery.NewDocumentFromReader(resp.Body)
	if err != nil {
		panic("parse fail")
	}

	// 클래스가 domain인 요소 찾기
	text := doc.Find(fmt.Sprintf("p#%s", domain)).Text()
	if text == "" {
		panic("text find fail")
	}

	return text
}

// http~/ + name + .num 를 path로 바이너리 생성
func Download(url string, name string, num int, path string) {
	// URL이 끝에 슬래시가 없으면 추가
	if url[len(url)-1] != '/' {
		url = url + "/"
	}

	// 파일 생성
	file, err := os.Create(path)
	if err != nil {
		panic("file creation fail")
	}
	defer file.Close()

	// 파일에 데이터 쓰기
	for i := 0; i < num; i++ {
		resp, err := http.Get(fmt.Sprintf("%s%s.%d", url, name, i))
		if err != nil {
			panic("download fail")
		}
		defer resp.Body.Close()

		if resp.StatusCode != http.StatusOK {
			panic("connection fail")
		}

		_, err = io.Copy(file, resp.Body)
		if err != nil {
			panic("writing fail")
		}
	}
}

// edgedriver 홈페이지에서 정보 받아오기, 홈페이지 + 행동정보 + zip경로 -> 버전 정수
func Driver(url string, datas []string, path string) int {
	// HTTP GET 요청
	resp, err := http.Get(url)
	if err != nil {
		panic("get fail")
	}
	defer resp.Body.Close()

	// HTTP 응답이 200 OK인지 확인
	if resp.StatusCode != http.StatusOK {
		panic("200 fail")
	}

	// HTML 파싱
	doc, err := goquery.NewDocumentFromReader(resp.Body)
	if err != nil {
		panic("parse fail")
	}

	// stable box 찾기
	var box *goquery.Selection
	doc.Find(datas[0] + "." + datas[1]).Each(func(i int, s *goquery.Selection) {
		text := strings.ToLower(strings.ReplaceAll(s.Text(), " ", ""))
		if strings.Contains(text, "stable") {
			box = s
		}
	})
	if box == nil {
		panic("No Stable WD")
	}

	// x64 button 찾기
	var but *goquery.Selection
	box.Find(datas[2] + "." + datas[3]).Each(func(i int, s *goquery.Selection) {
		text := strings.ToLower(strings.ReplaceAll(s.Text(), " ", ""))
		if strings.Contains(text, "x64") {
			but = s
		}
	})
	if but == nil {
		panic("No x64 WD")
	}

	// wd.zip link, ver int
	link, _ := but.Attr("href")
	ver := 0
	text := box.Find(datas[4] + "." + datas[5]).Text()
	temp := ""
	nums := []rune{'0', '1', '2', '3', '4', '5', '6', '7', '8', '9'}
	for _, r := range text {
		for _, l := range nums {
			if r == l {
				temp = temp + string(r)
			}
		}
		if r == '.' {
			ver, _ = strconv.Atoi(temp)
			break
		}
	}

	// wd.zip download
	file, err := os.Create(path)
	if err != nil {
		panic("file creation fail")
	}
	defer file.Close()
	resp, err = http.Get(link)
	if err != nil {
		panic("download fail")
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		panic("connection fail")
	}

	_, err = io.Copy(file, resp.Body)
	if err != nil {
		panic("writing fail")
	}

	return ver
}
